### PR TITLE
Cfy 4888 winrm support

### PR DESCRIPTION
**Pull request by Havivw**

### CODE REVIEW

File: plugin.yaml [hunk 1]
```diff
@@ -7,3 +7,8 @@ plugins:
         source: https://github.com/cloudify-cosmo/cloudify-fabric-plugin/archive/1.4.zip
         package_name: cloudify-fabric-plugin
         package_version: '1.4'
+    winrm:
+        executor: central_deployment_agent
+        source: https://github.com/cloudify-cosmo/cloudify-fabric-plugin/archive/1.4.zip
+        package_name: cloudify-fabric-plugin
+        package_version: '1.4'
```
Comment (EarthmanT) [inline plugin.yaml]: add a single extra line at the end of this file

File: winrm_plugin/README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# cloudify-winrm-plugin
+
+preliminary requirements:
+
+ubuntu side:
+sudo pip install pywinrm
```
Comment (EarthmanT) [inline winrm_plugin/README.md]: When you put pywinrm in the setup.py, you don't need to do pip install. Pip will automatically resolve that when this plugin is installed.

File: winrm_plugin/VARS_IN_BP [hunk 3]
```diff
@@ -0,0 +1,9 @@
+address: 192.168.11.36
+username: administrator
+password: Aa123456!
+local_file_path: ./example-script/script.py
+process: python
+delete_after_running: True
+remote_script_path: C:\Users\Administrator\Desktop
+port: 5985
+protocol: http
```
Comment (EarthmanT) [inline winrm_plugin/VARS_IN_BP]: Add line at the end.
rename to inputs.yaml

File: winrm_plugin/__init__.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+########
+# Copyright (c) 2014 GigaSpaces Technologies Ltd. All rights reserved
```
Comment (EarthmanT) [inline winrm_plugin/__init__.py]: 2016

File: winrm_plugin/example-script/script.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import sys
+
+
+def write():
+    try:
+        file = open('c:\\users\\administrator\\desktop\\python-succes.txt',
```
Comment (EarthmanT) [inline winrm_plugin/example-script/script.py]: Use this:

http://preshing.com/20110920/the-python-with-statement-by-example/

File: winrm_plugin/example-script/script.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+                    'a')
+        file.close()
+
+    except:
+        print('Something went wrong! Can\'t tell what?')
```
Comment (EarthmanT) [inline winrm_plugin/example-script/script.py]: Use this:

http://stackoverflow.com/questions/2052390/manually-raising-throwing-an-exception-in-python

File: winrm_plugin/tasks.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+from cloudify import ctx
+from cloudify.decorators import operation
+
+
+@operation
+def run_script(address, username, password, process, local_file_path,
+               delete_after_running=True, remote_script_path=None, port=5985,
+               protocol="http", **kwargs):
+
+    remote_script_file_name = "\\temp" + os.path.splitext(local_file_path)[1]
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: Use this:

https://docs.python.org/2/library/tempfile.html#tempfile.tempdir

File: winrm_plugin/tasks.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+def define_script_path(remote_script_path, is_cmd=True):
+
+    tmp_env_var = '%TEMP%' if is_cmd else '$env:TEMP'
+    return remote_script_path if ctx.node.properties['remote_script_path'] \
+        else tmp_env_var
+
+
+def get_conn(protocol, address, password, username, port):
+
+    endpoint = '{0}://{1}:{2}/wsman'.format(protocol, address, port)
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: When you write tests, make sure that you test this.

File: winrm_plugin/tasks.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    return winrm.Protocol(endpoint=endpoint, transport="plaintext",
+                          username=username, password=password)
+
+
+def get_remote_shell_id(conn):
+    return conn.open_shell()
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: Does this ever raise an exception.

File: winrm_plugin/tasks.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    tmp_env_var = '%TEMP%' if is_cmd else '$env:TEMP'
+    return remote_script_path if ctx.node.properties['remote_script_path'] \
+        else tmp_env_var
+
+
+def get_conn(protocol, address, password, username, port):
+
+    endpoint = '{0}://{1}:{2}/wsman'.format(protocol, address, port)
+    return winrm.Protocol(endpoint=endpoint, transport="plaintext",
+                          username=username, password=password)
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: Does this ever raise an exception?

File: winrm_plugin/tasks.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        encode_command = create_encoded_command(command)
+        ctx.logger.info('running command: {0}'.format(encode_command))
+        run_remote_command(remote_shell_id, process,  encode_command, " ", conn)
+
+
+def define_script_path(remote_script_path, is_cmd=True):
+
+    tmp_env_var = '%TEMP%' if is_cmd else '$env:TEMP'
+    return remote_script_path if ctx.node.properties['remote_script_path'] \
+        else tmp_env_var
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: You can just do:
```
return remote_script_path if remote_script_path else tmp_env_var
```

File: winrm_plugin/tasks.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+    for command in commands:
+        encode_command = create_encoded_command(command)
+        ctx.logger.info('running command: {0}'.format(encode_command))
+        run_remote_command(remote_shell_id, process,  encode_command, " ", conn)
+
+
+def define_script_path(remote_script_path, is_cmd=True):
+
+    tmp_env_var = '%TEMP%' if is_cmd else '$env:TEMP'
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: Consider using this:

https://docs.python.org/2/library/tempfile.html#tempfile.tempdir

File: winrm_plugin/tasks.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+def get_remote_shell_id(conn):
+    return conn.open_shell()
+
+
+def create_script_creation_command(local_file_path, powershell_path,
+                                   remote_script_file_name):
+
+    with open(local_file_path, 'r') as script_file:
+        script_content = script_file.read()
+    script_creator_cmd_prefix = '''$stream = [System.IO.StreamWriter] "{0}{1}";
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: What does this do?

File: winrm_plugin/tasks.py [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+                                   $s = @'\n'''.format(powershell_path,
+                                                       remote_script_file_name)
+
+    script_creator_cmd_suffix = \
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: What does this do?

File: winrm_plugin/tasks.py [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    conn = get_conn(protocol, address, password, username, port)
+
+    remote_shell_id = get_remote_shell_id(conn)
+
+    powershell_path = define_script_path(remote_script_path, False)
+
+    encoded_script = create_script_creation_command(
+            local_file_path, powershell_path, remote_script_file_name)
+
+    # copy the script file
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: move these here:
```
conn = get_conn(protocol, address, password, username, port)
remote_shell_id = get_remote_shell_id(conn)
```

File: winrm_plugin/tasks.py [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+    ctx.logger.info('Create script file on remote machine')
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: Maybe "Copying script file..."

File: winrm_plugin/tasks.py [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+    run_remote_command(remote_shell_id, 'powershell', '-encodedcommand',
+                       ' ' + encoded_script, conn)
+
+    cmd_path = define_script_path(remote_script_path)
+    ctx.logger.info('Runing the script on remote machine')
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: correct "Running"

File: winrm_plugin/tasks.py [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+    run_remote_command(remote_shell_id, process, cmd_path,
+                       remote_script_file_name, conn)
+
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: Maybe there should be some kind of validation that the script ran?

File: winrm_plugin/tasks.py [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+                 process, port=5985, protocol="http", **kwargs):
+
+    conn = get_conn(protocol, address, password, username, port)
+
+    remote_shell_id = get_remote_shell_id(conn)
+
+    for command in commands:
+        encode_command = create_encoded_command(command)
+        ctx.logger.info('running command: {0}'.format(encode_command))
+        run_remote_command(remote_shell_id, process,  encode_command, " ", conn)
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: single quotes

File: winrm_plugin/tasks.py [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+
+    if delete_after_running:
+        ctx.logger.info('Removing script file from remote machine')
+        run_remote_command(remote_shell_id, 'del', cmd_path,
+                           remote_script_file_name, conn)
+
+
+@operation
+def run_commands(commands, address, username, password,
+                 process, port=5985, protocol="http", **kwargs):
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: single quotes

File: winrm_plugin/tasks.py [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+
+    powershell_path = define_script_path(remote_script_path, False)
+
+    encoded_script = create_script_creation_command(
+            local_file_path, powershell_path, remote_script_file_name)
+
+    # copy the script file
+    ctx.logger.info('Create script file on remote machine')
+    run_remote_command(remote_shell_id, 'powershell', '-encodedcommand',
+                       ' ' + encoded_script, conn)
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: use:
' {0}'.format(encoded_script)

Comment (EarthmanT) [inline winrm_plugin/tasks.py]: Also pick/squash (come over and I'll explain when you get there)

File: winrm_plugin/tasks.py [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+
+    endpoint = '{0}://{1}:{2}/wsman'.format(winrm_protocol, address, winrm_port)
+    return winrm.Protocol(endpoint=endpoint, transport="plaintext",
+                          username=username, password=password)
+
+
+def get_remote_shell_id(conn):
+    try:
+        return conn.open_shell()
+    except winrm.exceptions.UnauthorizedError as remote_shell_error:
```
Comment (EarthmanT) [inline winrm_plugin/tasks.py]: change to 
```
except UnauthorizedError as e:
...
```

and then 
```
from winrm.exceptions import UnauthorizedError
```

**Comment by nir0s** [comment]
Closing as this isn't ready for a PR

**Pull request closed without merging**